BOM-2932: Replace Travis with GitHub CI

UsamaSadiq opened the pull request:
**Issue:** [BOM-2932](https://openedx.atlassian.net/browse/BOM-2932)

## Description
- Replace `Travis` with `GitHub CI`.

UsamaSadiq committed:
build: Replace Travis with GitHub CI

brianhw commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,56 @@
+          - python-version: 3.8
+            toxenv: quality
+          - python-version: 3.8
+            toxenv: docs
```
Do we know that `quality` and `docs` do not run in Python 3.8?   Is it worth enabling them if they pass?

pwnage101 replied:
IMO, it seems more straightforward to me to define an include list instead of an exclude list.  Imagine how much more complicated this gets when we introduce a 3rd python version during a transitional period...

pwnage101 replied:
My recommendation is to use tox-gh-actions as in here: https://github.com/edx/analytics-tools/blob/3b0f77ce885008ae3dfce27125971daae7e71bd7/.github/workflows/ci.yml

Define two jobs:  one to run python tests, another to run quality.  Now, your codecov step only needs to check matrix.python-version, and you do not need to define an include OR an exclude list at all!

UsamaSadiq replied:
`quality` and `docs` are passing with `Python 3.8`. We can implement `tox-gh-actions` later on since `Python 3.6` is soon going to be dropped so we won't need to create a matrix for `Python` versions.

brianhw commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,56 @@
+      run: tox
+
+    - name: Run Coverage
+      if: matrix.python-version == '3.6' && matrix.toxenv=='py36'
```
...likewise, do we want to run this for py3.8 too?

brianhw commented on README.rst:
```diff
@@ -83,9 +83,9 @@ For more information about these options, see the `Getting Help`_ page.
-    :target: https://travis-ci.com/edx/dbt-schema-builder
-    :alt: Travis
+.. |ci-badge| image:: https://github.com/edx/dbt-schema-builder/workflows/Python%20CI/badge.svg?branch=master
+        :target: https://github.com/edx/dbt-schema-builder/actions?query=workflow%3A%22Python+CI%22
```
nit: indentation is different for this...

brianhw commented on requirements/dev.in:
```diff
@@ -3,7 +3,7 @@
 -r pip-tools.txt          # pip-tools and its dependencies, for managing requirements files
 -r quality.txt            # Core and quality check dependencies
--r travis.txt             # tox and related dependencies
+-r ci.txt             # tox and related dependencies
```
nit: indentation of comment?

brianhw reviewed (approved):
This looks good -- no blockers.  I understand that some stuff hadn't been done in py3.8 before, and you might not want to include it here.   But I was just curious.   Indentation fixes are completely optional.

pwnage101 reviewed (changes requested):
While we're in here, I'd like to get it as good as possible and use tox-gh-actions

estute commented:
@pwnage101 are we doing that across the org? I'm less familiar with GH Actions, but I did a quick check in some of the other repos that we have migrated from Travis CI to Github Actions, and I did not see that library used.

estute commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,52 @@
+    - name: Install dependencies
+      run: pip install -r requirements/ci.txt
+
+    - name: Run py36 Tests
```
This name is misleading

UsamaSadiq replied:
Updated.

estute reviewed (commented):
The versioning in this repo feels weird to me. The setup.py file recognizes python3.6 through 3.8 as acceptable. The CI script runs tests for both versions, but only generates coverage metrics for 3.6. The deployment runs only for python3.8. However, the ONLY known usage of the code in this repo is here, running on python 3.7! https://github.com/edx/jenkins-job-dsl/blob/master/dataeng/jobs/analytics/SnowflakeSchemaBuilder.groovy#L48. This PR might not be the right place to address all of this, but I think it makes sense to have the tests, coverage and PyPI packaging running with the same version.

UsamaSadiq commented:
@estute I've dropped `Python<3.8` support and updated the dependencies accordingly.  Since `Python 3.6` is soon going to be deprecated. `Ends in 1 month and 1 week (23 Dec 2021)`. 
Removing `Python<3.8` also simplifies the checks for CI.

estute commented:
@jazibhumayun are we ok merging this? I know that you have been doing some work to get this running on python 3.8 in production.

jazibhumayun commented:
> @jazibhumayun are we ok merging this? I know that you have been doing some work to get this running on python 3.8 in production.

If this PR can stay a little longer I would prefer merging my two PRs first, as I have resolved dependencies recently using python 3.8 and the jenkins job schema builder job that runs is still using python 3.7

[https://github.com/edx/dbt-schema-builder/pull/53](https://github.com/edx/dbt-schema-builder/pull/53)
[https://github.com/edx/jenkins-job-dsl/pull/1474](https://github.com/edx/jenkins-job-dsl/pull/1474)

estute commented:
@UsamaSadiq the dbt upgrade is complete. We can move forward with this PR. I have created a branch based on your PR to rebase `main` onto it here (https://github.com/edx/dbt-schema-builder/pull/54), and will work on getting the tests passing tomorrow.

UsamaSadiq committed:
build: set the default main branch for push trigger

jazibhumayun commented:
I reviewed the PR and checked tests and PyPI packaging running with the same version (python 3.8).
Jenkins job that runs schema builder is also using python 3.8 [https://github.com/edx/jenkins-job-dsl/blob/master/dataeng/jobs/analytics/SnowflakeSchemaBuilder.groovy#L48](https://github.com/edx/jenkins-job-dsl/blob/master/dataeng/jobs/analytics/SnowflakeSchemaBuilder.groovy#L48)

As SRE/BOM wants to discontinue the usage of Travis so I will create a ticket to address @pwnage101's comments on this [PR](https://github.com/edx/dbt-schema-builder/pull/54) and will address those changes in a separate PR

jazibhumayun commented:
I have created this [ticket](https://openedx.atlassian.net/browse/DENG-1099) to address to Troy's comments
